File: MK-52_ESP32/src/Receiver.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receiver.hpp"

using namespace MK52_Interpreter;

//
// Inits generic receiver
//
Receiver::Receiver( void *components[]) {
    _components = components;
    _kbd = (MK52_Hardware::KBD_Manager *)components[COMPONENT_KBD_MANAGER];
    _lcd = (MK52_Hardware::LCD_Manager *)components[COMPONENT_LCD_MANAGER];
    _rpnf = (RPN_Functions*)components[COMPONENT_FUNCTIONS];
    memset( _funlabel, ' ', 3);
    _funlabel[3] = 0;
}

void Receiver::activate( int8_t prevReceiver){
    _return_to = prevReceiver;
}

uint8_t Receiver::tick( uint8_t scancode){
    return 0;
}

char Receiver::_convertButton(const char *list, uint8_t scancode){
    char c = (char)pgm_read_byte(list + scancode);
    if( c == '#') return 0;
    return c;
}

File: MK-52_Simulator/Receivers/Receiver.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    //
    // Implements a generic empty receiver
    //
    public class Receiver
    {
        public string Moniker = "Undefined";

        protected MK52_Host _parent;
        protected uint _mode = 0;
        protected string _return_to = "";
        protected StringBuilder _text = new StringBuilder();
        protected UniversalValue _uv = new UniversalValue();

        public Receiver(MK52_Host parent)
        {
            _parent = parent;
        }

        public virtual void activate( string prevReceiver)
        {
            this._return_to = prevReceiver;
        }

        public virtual byte tick( byte scancode)
        {
            return (byte)0;
        }

        public virtual string toString()
        {
            return _text.ToString();
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("Receiver [");
            sb.Append(Moniker);
            sb.Append("]: ");
            sb.Append(_text.ToString());
            return sb.ToString();
        }

        public virtual string toTrimmedString()
        {
            return _text.ToString().TrimStart();
        }

        public string getReturnReceiverMoniker()
        {
            if (_return_to.Length <= 0) return "AUTO_N"; // as minimum need to return to AUTO_N
            return _return_to;
        }

        protected char _convertButton( string list, uint scancode)
        {
            if (scancode >= list.Length) return (char)0;
            char c = list[(int)scancode];
            if( c == '#') return (char)0;
            return c;
        }
    }
}
